### PR TITLE
Helm install script

### PR DESCRIPTION
**Pull request by Fortune-Ndlovu**
Issue: [RHIDP-2706](https://issues.redhat.com/browse/RHIDP-2706) - script CLI helm installation for OCP

Continued from [RHIDP-2614](https://issues.redhat.com/browse/RHIDP-2614) re: https://github.com/redhat-developer/red-hat-developers-documentation-rhdh/pull/287

Instead of giving people 20 steps to iterate though, AND having the same steps captured by QE when doing install validation tests, we should instead:

- I create a script that can handle both install paths
- Link to that from the docs
- Reuse it in testing

The script does the following:

- detect if oc is installed; fail if not found and report error to user with call to action re: installing it
- detect if user is logged into a cluster and fail if not with call to action (oc command to log in)
- if cluster router base can be detected, use that value to update the helm chart installation; if not, fail and request user pass in commandline flag as cluster router base could not be detected (can test this failure case with dev sandbox; can test passing case with clusterbot)

The script is stored in https://github.com/redhat-developer/rhdh-chart/ using a similar documentation/script approach to

- https://github.com/janus-idp/operator/tree/main/.rhdh/docs
- https://github.com/janus-idp/operator/tree/main/.rhdh/scripts

- I have also created a detailed document with more instructions as to how to use the script as an install.md file also within the scripts directory.
 
## Checklist

- [ ] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the `values.yaml` and added to the README.md. The [pre-commit](https://pre-commit.com/) utility can be used to generate the necessary content. Use `pre-commit run -a` to apply changes.
- [ ] JSON Schema template updated and re-generated the raw schema via `pre-commit` hook.
- [ ] List tests pass for Chart using the [Chart Testing](https://github.com/helm/chart-testing) tool and the `ct lint` command.

### CODE REVIEW

**Comment by coreydaley** [comment]
The name of the script `install-helm.sh` makes it sound like it installs the Helm executable. I think it should just be named `install.sh`.

File: scripts/install-helm.sh [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+# Check if either oc or kubectl is installed
+if command_exists oc; then
+    CLI="oc"
+elif command_exists kubectl; then
+    CLI="kubectl"
+else
+    echo "Error: Neither 'oc' nor 'kubectl' is installed. Please install one of them to proceed."
+    exit 1
+fi
```
Comment (coreydaley) [inline scripts/install-helm.sh:52]: We should provide a flag that can be used to override the automatic detection here in case someone wants to use kubectl instead of oc even if they are both installed.

File: scripts/install-helm.sh [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    echo "Error: Cluster router base could not be detected. Please provide it using the --router-base flag."
+    exit 1
+fi
+
+# Parse command-line arguments for optional router base
+while [[ "$#" -gt 0 ]]; do
+    case $1 in
+        --router-base) ROUTER_BASE="$2"; shift ;;
+        *) echo "Unknown parameter passed: $1"; exit 1 ;;
+    esac
```
Comment (coreydaley) [inline scripts/install-helm.sh:83]: This would be more readable (especially by `bash` noobs) if the commands were on their own lines, I don't see a reason to be worried about the length of the script and make the lines harder to read (this isn't ruby!)

File: scripts/install-helm.sh [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+detect_cluster_router_base() {
+    if [ "$CLI" == "oc" ]; then
+        ROUTER_BASE=$($CLI get route -n openshift-console -o=jsonpath='{.items[0].spec.host}')
+    else
+        ROUTER_BASE=$($CLI get ingress -n default -o=jsonpath='{.items[0].spec.rules[0].host}')
+    fi
+}
+
+# Detect cluster router base
+detect_cluster_router_base
```
Comment (coreydaley) [inline scripts/install-helm.sh:70]: This function doesn't seem to be used anywhere else, probably not a need for the complexity.

File: scripts/install-helm.sh [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    exit 1
+fi
+sed -i "s|routerBase:.*|routerBase: $ROUTER_BASE|" "$VALUES_FILE"
+
+# Define the path to the Helm chart directory
+HELM_CHART_DIR="../charts/backstage"
+if [ ! -d "$HELM_CHART_DIR" ]; then
+    echo "Error: Helm chart directory not found at $HELM_CHART_DIR"
+    exit 1
+fi
```
Comment (coreydaley) [inline scripts/install-helm.sh:102]: This should probably be at the top of the file near the check for `oc` and `kubectl`, along with checking for the `values.yaml` file, if those don't exist, there is no need to run the rest of the script.

File: scripts/install-helm.sh [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+# Proceed with Helm installation
+helm install my-release "$HELM_CHART_DIR" --values "$VALUES_FILE"
```
Comment (coreydaley) [inline scripts/install-helm.sh:105]: People will probably want to choose their own name instead of `my-release`, i think we also support a `--generate-name` command with the helm install? Should support that also if that is the case.

**Review by coreydaley — CHANGES_REQUESTED** [review]
Off to a good start! I made some suggestions.

File: scripts/install-helm.sh (hunk 1 above)
Comment (nickboldt) [reply scripts/install-helm.sh:52]: why?

File: scripts/install-helm.sh (hunk 5 above)
Comment (nickboldt) [reply scripts/install-helm.sh:105]: agreed, the namespace could be autodetected if not passing in via commandline

File: scripts/install-helm.sh (hunk 1 above)
Comment (Fortune-Ndlovu) [reply scripts/install-helm.sh:52]: Great Idea -  For this I have added an option cli that takes either oc or kubectl as a value. The script uses the specified cli tool if provided; otherwise, it will fall back to the automatic detection.

File: scripts/install-helm.sh (hunk 3 above)
Comment (Fortune-Ndlovu) [reply scripts/install-helm.sh]: With the new suggested changes, I use it to detect cluster router base if not provided.

File: scripts/install-helm.sh (hunk 5 above)
Comment (Fortune-Ndlovu) [reply scripts/install-helm.sh]: Sounds good - I have added user-defined release names and support for the `--generate-name` option.

File: scripts/install.sh [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    fi
+fi
+
+# Update Helm chart with the detected or provided router base
+echo "Using router base: $ROUTER_BASE"
+sed -i "s|routerBase:.*|routerBase: $ROUTER_BASE|" "$VALUES_FILE"
+
+# Construct Helm install command
+HELM_CMD="helm install"
+if $GENERATE_NAME; then
```
Comment (nickboldt) [inline scripts/install.sh:168]:
```suggestion
if [[ $GENERATE_NAME ]]; then
```

File: scripts/install.sh [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    HELM_CMD+=" --generate-name"
+else
```
Comment (nickboldt) [inline scripts/install.sh:170]: this could be an elif, since you only have one else. would remove one level of nesting for easier reading.

File: scripts/install.sh [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    if [[ -z "$RELEASE_NAME" ]]; then
+        echo "Error: Either --release-name must be specified or --generate-name must be used."
```
Comment (nickboldt) [inline scripts/install.sh:172]:
```suggestion
        echo "Error: Only one of --release-name or --generate-name can be used."
```

File: scripts/install.sh [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+# If detection fails, prompt user for input
+if [[ -z "$ROUTER_BASE" ]]; then
+    echo "Error: Cluster router base could not be detected. Please provide it using the --router-base flag."
+    exit 1
+fi
+
+# Detect namespace if not provided
+if [[ -z "$NAMESPACE" ]]; then
+    NAMESPACE=$($CLI config view --minify --output 'jsonpath={..namespace}')
```
Comment (nickboldt) [inline scripts/install.sh:155]: Are you sure you want `..namespace`, not `.namespace` ?

```suggestion
    NAMESPACE=$($CLI config view --minify --output 'jsonpath={..namespace}')
```

File: scripts/install.sh [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    fi
+    HELM_CMD+=" $RELEASE_NAME"
+fi
+
+HELM_CMD+=" $HELM_CHART_DIR --values $VALUES_FILE --namespace $NAMESPACE"
+
+# Execute Helm install command
+eval $HELM_CMD
+
+echo "Helm installation completed successfully."
```
Comment (nickboldt) [inline scripts/install.sh:183]: What if the above command failed?

```suggestion
if [[ $? -eq 0 ]]; then 
    echo "Helm installation completed successfully."
else
    echo "Something went wrong with helm installation!"; exit 1
fi
```

File: scripts/install.sh [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+# Script to handle CLI helm installation for k8s and OCP
+#
+# Requires: oc or kubectl
```
Comment (nickboldt) [inline scripts/install.sh:18]:
```suggestion
# Requires: oc or kubectl and an active login sesstion to a cluster. 
```

File: scripts/install-helm.sh (hunk 1 above)
Comment (nickboldt) [reply scripts/install-helm.sh]: yeah the "why?" was at Corey because I can't grok why if you have both tools installed you'd care which one was used to perform the cluster introspection.

File: scripts/install.sh [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+It detects whether 'oc' or 'kubectl' is installed and ensures that the user is logged into a cluster.
+The script also attempts to detect the cluster router base and updates the Helm chart configuration accordingly.
+
+Usage:
+  $0 [OPTIONS]
+
+Options:
+  --cli <oc|kubectl>          : Specify the CLI tool to use (overrides auto-detection).
+  --router-base <router-base> : Manually provide the cluster router base if auto-detection fails.
+  --release-name <name>       : Specify a custom release name for the Helm chart.
```
Comment (nickboldt) [inline scripts/install.sh:32]: In my helm-for-ocp script, I hardcode name to `redhat-developer-hub` and then the namespace defaults to `namespace="rhdh-helm"`

https://github.com/rhdh-bot/openshift-helm-charts/blob/rhdh-1-rhel-9/installation/install.sh

so we end up with this route: https://redhat-developer-hub-${namespace}.${CLUSTER_ROUTER_BASE}

Here we could get all manner of options using <name>.<namespace>.<router-base>.

I wonder if there's value in setting defaults in case people fail to pass in overrides so we can default to things we know work...

Comment (coreydaley) [reply scripts/install.sh:32]: It should never matter what name and namespace are used, everything should be routed correctly automatically. If someone wants to install multiple developer hub instances on the same cluster, they would currently have to go in and modify the script if we hard code the values, it would make more sense to have flags that they can use to provide the name and namespace I think.

File: scripts/install.sh [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+# Check if the user is logged into a cluster
+if ! $CLI whoami >/dev/null 2>&1; then
+    echo "Error: You are not logged into a cluster. Please log in using '$CLI login' or '$CLI config set-context'."
+    exit 1
+fi
+
+# Function to detect cluster router base
+detect_cluster_router_base() {
+    if [ "$CLI" == "oc" ]; then
+        ROUTER_BASE=$($CLI get route -n openshift-console -o=jsonpath='{.items[0].spec.host}')
```
Comment (coreydaley) [inline scripts/install.sh:136]: Will this command work if the user is not a cluster administrator?

File: scripts/install.sh (hunk 9 above)
Comment (coreydaley) [reply scripts/install.sh:155]: @nickboldt did you forget to make the change in your suggested change? :)

File: scripts/install.sh [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+    NAMESPACE=$($CLI config view --minify --output 'jsonpath={..namespace}')
+    if [[ -z "$NAMESPACE" ]]; then
+        echo "Error: Namespace could not be detected. Please provide it using the --namespace flag."
+        exit 1
+    fi
+fi
+
+# Update Helm chart with the detected or provided router base
+echo "Using router base: $ROUTER_BASE"
+sed -i "s|routerBase:.*|routerBase: $ROUTER_BASE|" "$VALUES_FILE"
```
Comment (coreydaley) [inline scripts/install.sh:164]: I haven't tried it out, but i feel like changing the contents of the values file is not a good choice and we should use the `--set` or `--set-string` option to pass the clusterRouterBase instead when the command runs.

File: scripts/install.sh [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+# Construct Helm install command
+HELM_CMD="helm install"
```
Comment (rm3l) [inline scripts/install.sh:167]:
```suggestion
HELM_CMD="helm upgrade --install"
```

Should we attempt to upgrade an existing release, with the option to install if the release does not exist yet?

File: scripts/install.sh (hunk 13 above)
Comment (rm3l) [inline scripts/install.sh:136]: This would return the Route URL of the OpenShift Web Console, and I'm not sure this is what we want:
```
$ ./install.sh
Using router base: console-openshift-console.apps.ci-ln-cd29yg2-76ef8.origin-ci-int-aws.dev.rhcloud.com
^C
```
In the example above, what we call cluster router base is `apps.ci-ln-cd29yg2-76ef8.origin-ci-int-aws.dev.rhcloud.com`.

I usually use the following command to get the cluster router base domain on OCP. 
```
oc get ingress.config.openshift.io/cluster -oyaml | yq '.spec.domain'
```

This adds a dependency on `yq` to extract data from the YAML output, but feel free to use `-o=jsonpath=...` to extract from the JSON output if we don't want that dependency. Left as an exercise to you ;)

NOTE: this requires access to the cluster-wide Ingress Config on the cluster, and won't work on Dev Sandbox for example.

File: scripts/install.sh [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+    echo "Error: You are not logged into a cluster. Please log in using '$CLI login' or '$CLI config set-context'."
+    exit 1
+fi
+
+# Function to detect cluster router base
+detect_cluster_router_base() {
+    if [ "$CLI" == "oc" ]; then
+        ROUTER_BASE=$($CLI get route -n openshift-console -o=jsonpath='{.items[0].spec.host}')
+    else
+        ROUTER_BASE=$($CLI get ingress -n default -o=jsonpath='{.items[0].spec.rules[0].host}')
```
Comment (rm3l) [inline scripts/install.sh:138]: I think this requires a bit more caution for standard K8s clusters. Here, you are getting the Ingress resources from the _default_ namespace, but I'm not sure this makes sense..
The upstream chart exposes the ability to create an Ingress resource (in the same namespace where the Chart is deployed) by setting [`upstream.ingress.enabled`](https://github.com/backstage/charts/blob/main/charts/backstage/values.yaml#L59), and the RHDH Chart sets the ingress host to the [value of `global.host`](https://github.com/redhat-developer/rhdh-chart/blob/main/charts/backstage/values.yaml#L188).
So I'm wondering if detecting the cluster router base actually makes sense on non-OCP clusters.

**Review by rm3l — CHANGES_REQUESTED** [review]

File: scripts/install.sh [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+"
+}
+
+# Function to check if a command exists
+command_exists() {
+    command -v "$1" >/dev/null 2>&1
+}
+
+# Check if required files and directories exist
+HELM_CHART_DIR="../charts/backstage"
```
[the remark this answers is not in the document]
Comment (rm3l) [reply scripts/install.sh:52]: Another question: unless I misunderstood it, I thought the goal of this script was to ultimately be included in our product docs, so that users could call it instead of running several commands (against either OCP or K8s).
In this case, as it is right now, it would require people to clone this repo first. Is that what we want?
Should we add the chart repo instead?
Or will this be handled in a separate install script?

File: scripts/install.sh (hunk 13 above)
Comment (nickboldt) [reply scripts/install.sh:136]: easy to test using a dev sandbox instance https://developers.redhat.com/developer-sandbox/get-started

File: scripts/install.sh (hunk 17 above)
[the remark this answers is not in the document]
Comment (nickboldt) [reply scripts/install.sh:52]: hmm. good point.

so the prod docs would have:
* clone the repo
* run the script

instead of:
* run
* these
* ten
* individual
* steps

And QE could do the same when installing stuff...

**Comment by nickboldt** [comment]
Apologies but I've just realized two things:

a) this issue was only about helm CLI installs on OCP, and
b) we already have doc for aks and eks helm cli installs:

https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.1/html-single/administration_guide_for_red_hat_developer_hub/index#proc-rhdh-deploy-eks-using-helm_admin-rhdh and
https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.1/html-single/administration_guide_for_red_hat_developer_hub/index#proc-rhdh-deploy-aks-using-helm_admin-rhdh


so perhaps we can make this script simpler and just handle the OCP install case. Keep It Super Simple, as they say.

File: scripts/install.sh [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+    shift
+done
+
+# If CLI is not specified, detect automatically
+if [[ -z "$CLI" ]]; then
+    detect_cli
+fi
+
+# Check if the user is logged into a cluster
+if ! $CLI whoami >/dev/null 2>&1; then
```
[the remark this answers is not in the document]
Comment (Fortune-Ndlovu) [reply scripts/install.sh:129]: You are correct - Since the goal of this script is to ensure that users are prepared before they proceed with the installation, checking if the user is logged in might not be necessary because Helm and other commands will fail with appropriate error messages. I will remove this logic as it is not necessary.

File: scripts/install.sh (hunk 13 above)
Comment (Fortune-Ndlovu) [reply scripts/install.sh:131]: I see.. so I can update the detect_cluster_router_base function to use oc get ingress.config.openshift.io/cluster to fetch the correct router base domain on OpenShift
I think I can also handle the JSON path extraction directly without adding a dependency on yq

File: scripts/install.sh [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+    NAMESPACE=$(oc config view --minify --output 'jsonpath={..namespace}')
+else
+    echo "Error: Namespace could not be detected. Please provide it using the --namespace flag."
+    exit 1
+fi
+
+# Update Helm chart with the detected or provided router base if using default values file
+if [[ "$VALUES_FILE" == "$DEFAULT_VALUES_FILE" ]]; then
+    echo "Using router base: $ROUTER_BASE"
+    sed -i "s|routerBase:.*|routerBase: $ROUTER_BASE|" "$VALUES_FILE"
```
Comment (rm3l) [inline scripts/install.sh:131]: As commented out by Corey in https://github.com/redhat-developer/rhdh-chart/pull/27#discussion_r1646718603, let's not modify the base values.yaml file. Otherwise, since we have cloned the repo, this would result in an unnecessary diff.
You can use the `--set global.clusterRouterBase="$ROUTER_BASE"` to change the value of this parameter.

File: scripts/install.md [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+
+NAMESPACE: ...
+
+STATUS: ...
+
+REVISION: ...
+
+Helm installation completed successfully.
+
+You can run: `helm version` to confirm this installation.
```
Comment (rm3l) [inline scripts/install.md:95]:
```suggestion
You can run: `helm list --namespace $myProject` to confirm this installation.
```

`helm version` only returns the version of Helm. [`helm list`](https://helm.sh/docs/helm/helm_list/) (or `helm ls`) will list the installed releases.

File: scripts/install.md [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+    **Specify Custom Values File** - To use a custom values file for the Helm chart: \
+   `./install.sh --values /path/to/values.yaml`
+
+
+
+
+## **Troubleshooting**
+
+
+* Missing <code>oc</code> - </strong> If the script outputs an error indicating that neither <code>oc</code> nor <code>kubectl</code> is installed, please install the appropriate CLI tool for your cluster:
```
Comment (rm3l) [inline scripts/install.md:60]: I guess this bullet point is no longer necessary since the script only picks `oc`.

File: scripts/install.sh [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+    shift
+done
+
+# Function to detect cluster router base
+detect_cluster_router_base() {
+    ROUTER_BASE=$(oc get ingress.config.openshift.io/cluster -o=jsonpath='{.spec.domain}')
+}
+
+# Detect cluster router base if not provided
+if [[ -z "$ROUTER_BASE" ]]; then
```
Comment (rm3l) [inline scripts/install.sh:113]: Just tried this on [Dev Sandbox](https://developers.redhat.com/developer-sandbox), and I was expecting to see the `Error: Cluster router base could not be detected. Please provide it using the --router-base flag.` message, but got a generic `oc` error instead:

```
$  ./scripts/install.sh --generate-name
Error from server (Forbidden): ingresses.config.openshift.io "cluster" is forbidden: User "rh-ee-asoro" cannot get resource "ingresses" in API group "config.openshift.io" at the cluster scope
```

Even when I passed the `--router-base` flag, I got an error:

```
$ ./scripts/install.sh --generate-name --router-base apps.sandbox-m4.g2pi.p1.openshiftapps.com
Error: Cluster router base could not be detected. Please provide it using the --router-base flag.
```

Am I missing something?

**Review by rm3l — CHANGES_REQUESTED** [review]

**Comment by Fortune-Ndlovu** [comment]
@coreydaley @nickboldt @rm3l I have updated the script can you guys take another look please.

File: scripts/install.sh [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+    NAMESPACE=$(oc config view --minify --output 'jsonpath={..namespace}')
+    if [[ -z $NAMESPACE ]]; then
+        echo "Error: Namespace could not be detected. Please provide it using the --namespace flag."
+        exit 1
+    fi
+fi
+
+# Update Helm chart with the detected or provided router base if using default values file
+if [[ "$VALUES_FILE" == "$DEFAULT_VALUES_FILE" ]]; then
+    EXTRA_HELM_ARGS+="--set global.clusterRouterBase=$ROUTER_BASE"
```
Comment (rm3l) [inline scripts/install.sh:132]:
```suggestion
    EXTRA_HELM_ARGS+=" --set global.clusterRouterBase=$ROUTER_BASE"
```

A space is needed, otherwise, there is an error with the resulting command:
```
$ ./scripts/install.sh --release-name rhdh --router-base apps.sandbox-m4.g2pi.p1.openshiftapps.com --set upstream.backstage.image.repository=rhdh/rhdh-hub-rhel9 --values .asoro/values-with-aws-auth.yaml
Executing: helm install rhdh ./scripts/../charts/backstage --namespace rh-ee-asoro-dev  --set upstream.backstage.image.repository=rhdh/rhdh-hub-rhel9 --values .asoro/values-with-aws-auth.yaml--set global.clusterRouterBase=apps.sandbox-m4.g2pi.p1.openshiftapps.com
Error: INSTALLATION FAILED: expected at most two arguments, unexpected arguments: global.clusterRouterBase=apps.sandbox-m4.g2pi.p1.openshiftapps.com
Something went wrong with Helm installation!
```

File: scripts/install.sh [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+if [[ $GENERATE_NAME == true ]]; then
+    HELM_CMD+=" --generate-name"
+elif [[ -z "$RELEASE_NAME" ]]; then
+    echo "Error: Either --release-name must be specified or --generate-name must be used."
+    exit 1
+else
+    HELM_CMD+=" $RELEASE_NAME"
+fi
+
+HELM_CMD+=" $HELM_CHART_DIR --values $VALUES_FILE --namespace $NAMESPACE $EXTRA_HELM_ARGS"
```
Comment (rm3l) [inline scripts/install.sh:146]:
```suggestion
HELM_CMD+=" $HELM_CHART_DIR --namespace $NAMESPACE $EXTRA_HELM_ARGS"
```

It is not needed to specify the path to the Chart default `values.yaml` file. If user explicitly passes `--values`, it will be captured in `$EXTRA_HELM_ARGS`

File: scripts/install.sh [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+done
+
+# Function to detect cluster router base
+detect_cluster_router_base() {
+    ROUTER_BASE=$(oc get ingress.config.openshift.io/cluster -o=jsonpath='{.spec.domain}')
+}
+
+# Detect cluster router base if not provided
+if [[ -z "$ROUTER_BASE" ]]; then
+    detect_cluster_router_base
```
[the remark this answers is not in the document]
Comment (Fortune-Ndlovu) [reply scripts/install.sh]: I agree!

File: scripts/install.sh [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+if [[ -z "$NAMESPACE" ]]; then
+    NAMESPACE=$(oc config view --minify --output 'jsonpath={..namespace}')
+    if [[ -z $NAMESPACE ]]; then
+        echo "Error: Namespace could not be detected. Please provide it using the --namespace flag."
+        exit 1
+    fi
+fi
+
+# Update Helm chart with the detected or provided router base if using default values file
+if [[ "$VALUES_FILE" == "$DEFAULT_VALUES_FILE" ]]; then
```
[the remark this answers is not in the document]
Comment (Fortune-Ndlovu) [reply scripts/install.sh:127]: You are right! since the `ROUTER_BASE` is always required and will either be provided by the user or auto-detected, there's no need to add the router base to the Helm arguments conditionally.

**Comment by Fortune-Ndlovu** [comment]
cc @rm3l 

for review, please

File: scripts/install.sh [hunk 27]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+# Function to check if a command exists
+command_exists() {
+    command -v "$1" >/dev/null 2>&1
+}
+
+# Check for required commands
+if ! command_exists helm || ! command_exists oc; then
+    echo "Error: Required commands 'helm' and 'oc' are not available."
```
Comment (nickboldt) [inline scripts/install.sh:54]: you're actually trapping for the condition where EITHER is missing. A better approach would be to trap for each one separately, and then provide a call to action for the user so they know where to get oc or helm to install it.


https://docs.openshift.com/container-platform/4.16/cli_reference/openshift_cli/getting-started-cli.html

thus you might do:

```
# Check for required commands
if ! command_exists helm; then
    echo "Error: 'helm' is required. Install it from https://docs.openshift.com/container-platform/4.16/applications/working_with_helm_charts/installing-helm.html to continue."
    exit 1
fi
if ! command_exists oc; then 
    echo "Error: Required 'oc' is required. Install it from https://docs.openshift.com/container-platform/4.16/cli_reference/openshift_cli/getting-started-cli.html to continue."
    exit 1
fi
```

File: scripts/install.md [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+
+NAMESPACE: ...
+
+STATUS: ...
+
+REVISION: ...
+
+Helm installation completed successfully.
+
+You can run: `helm list --namespace $myProject` to confirm this installation.
```
Comment (nickboldt) [inline scripts/install.md:77]: should we run this at the end of the install to verify installation worked?

File: scripts/install.md [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+## **Troubleshooting**
+
+* Missing <code>oc</code> - </strong> If the script outputs an error indicating <code>oc</code> is not installed, please install the CLI tool for your cluster:
+    * For OpenShift: Install <code>oc</code>
+
+* <strong>Not Logged Into Cluster: </strong> If you are not logged into a cluster, follow these steps to log in:
+    * For OpenShift:
+ `oc login <your-cluster-url>`
+
+* <strong>Router Base Detection Failed: - </strong> If the script cannot detect the cluster router base, manually provide it using the <code>--router-base</code> flag as shown in the example usage section.
```
Comment (nickboldt) [inline scripts/install.md:54]: this is already covered by an error message IN the script. Does it need to ALSO be in a readme?

File: scripts/install.md (hunk 28 above)
Comment (Fortune-Ndlovu) [reply scripts/install.md:77]: yep i would say running `helm list --namespace $myProject` at the end of the script serves as additional manual check for the user and we do already have a condition at the end of the script to verify the installation.

File: scripts/install.md (hunk 29 above)
Comment (Fortune-Ndlovu) [reply scripts/install.md:54]: I would say yes because it provides user awareness just in case users might not read the entire script or miss the error message when running it, also since its the md file it felt complete to cover common issues and troubleshooting steps even if it caused slight repetition.

File: scripts/install.md (hunk 28 above)
Comment (nickboldt) [reply scripts/install.md:77]: I guess this does align with the standard RH doc approach of "do a thing" then "verify it was thing'd correctly by running this other thing"

File: scripts/install.sh [hunk 30]
```diff
@@ -0,0 +1,150 @@
...
+HELM_CMD+=" $HELM_CHART_DIR --namespace $NAMESPACE $EXTRA_HELM_ARGS"
+
+# Execute Helm install command
+echo "Executing: $HELM_CMD"
+
+if eval "$HELM_CMD"; then
+    echo "Helm installation completed successfully."
+else
+    echo "Something went wrong with Helm installation!"
+    exit 1
```
Comment (nickboldt) [inline scripts/install.sh:149]:
```suggestion
    helm list --namespace $NAMESPACE
    exit 1
```

File: scripts/install.md (hunk 29 above)
Comment (nickboldt) [reply scripts/install.md:54]: true but "as shown in the example usage section points to the same usage and doesn't elaborate. So you have the same instructions 3 times now:
* example
* troubleshooting
* the script's output itself

This means if we have to update the code, you have to ALSO update the doc in 2 other places. You're making extra maintenance work for yourself here. 

And you haven't yet explained that `<your-router-base>` should be `apps.yourcluster-goes-here.tld`, as the example doesn't actually provide an example. 

Finally, there is no "example usage section" ... do you mean the **Installation Steps** section?

File: scripts/install.md [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+
+     ```shell
+      git clone --depth 1 https://github.com/redhat-developer/rhdh-chart.git
+      cd rhdh-charts/scripts
+    ```
+
+2. **Run the Script** - Execute the script to install the Helm chart. The script will automatically detect that you are using `oc` and ensure you are logged into your cluster. \
+`./install.sh`
+
+3. **Specify Router Base** - If the script cannot automatically detect the cluster router base, you can provide it manually using the `--router-base` flag, example:\
```
Comment (nickboldt) [inline scripts/install.md:30]: should items 3 - 7 be under a "Usage" section? This sequential numbering after steps 1 and 2 imply that you also have to run 5 more steps...

File: scripts/install.md [hunk 32]
```diff
@@ -0,0 +1,77 @@
...
+   `./install.sh --values /path/to/values.yaml`
+
+## **Troubleshooting**
+
+* Missing <code>oc</code> - </strong> If the script outputs an error indicating <code>oc</code> is not installed, please install the CLI tool for your cluster:
+    * For OpenShift: Install <code>oc</code>
+
+* <strong>Not Logged Into Cluster: </strong> If you are not logged into a cluster, follow these steps to log in:
+    * For OpenShift:
+ `oc login <your-cluster-url>`
```
Comment (nickboldt) [inline scripts/install.md:52]: better to link to the official guide https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-logging-in_cli-developer-commands which explains more than just the single commandline option... like passing in tokens and usernames and passwords...

File: scripts/install.md [hunk 33]
```diff
@@ -0,0 +1,77 @@
...
+
+6. **Specify Namespace** - To specify the namespace for the Helm release: \
+`./install.sh --namespace mynamespace`
+
+7. **Specify Custom Values File** - To use a custom values file for the Helm chart: \
+   `./install.sh --values /path/to/values.yaml`
+
+## **Troubleshooting**
+
+* Missing <code>oc</code> - </strong> If the script outputs an error indicating <code>oc</code> is not installed, please install the CLI tool for your cluster:
```
Comment (nickboldt) [inline scripts/install.md:47]: why only mention `oc` ? `helm` is also required. https://github.com/redhat-developer/rhdh-chart/pull/27/files#diff-e076e43cb817ecc8b3c2ed18cd58baa50544ecc66293ed6e61129c43ca6a09ccR53-R60

File: scripts/install.md [hunk 34]
```diff
@@ -0,0 +1,77 @@
...
+    * For OpenShift: Install <code>oc</code>
+
+* <strong>Not Logged Into Cluster: </strong> If you are not logged into a cluster, follow these steps to log in:
```
Comment (nickboldt) [inline scripts/install.md:50]: this is only one step. And this could be simplified to:

```suggestion
* <strong>Not Logged Into Cluster: </strong> To log in to your cluster using the `oc` command, see https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-logging-in_cli-developer-commands
```

File: scripts/install.md [hunk 35]
```diff
@@ -0,0 +1,77 @@
...
+    * For OpenShift:
+ `oc login <your-cluster-url>`
+
+* <strong>Router Base Detection Failed: - </strong> If the script cannot detect the cluster router base, manually provide it using the <code>--router-base</code> flag as shown in the example usage section.
+
+* <strong>Error: INSTALLATION FAILED:</strong>
+
+     An error occurred while checking for chart dependencies. You may need to run `helm dependency build` to fetch missing dependencies: found in Chart.yaml, but missing in charts/ directory: common, backstage:
```
Comment (nickboldt) [inline scripts/install.md:58]: if this can happen, why are we not just DOING it as part of the script?? Why leave it as a "haha gotcha" for the user?

File: scripts/install.md [hunk 36]
```diff
@@ -0,0 +1,77 @@
...
+  `helm dependency build`
+
+An example of the installation output should be:
+
+Using router base: example.com
```
Comment (nickboldt) [inline scripts/install.md:63]:
```suggestion
Using router base: apps.example.com
```

image

File: scripts/install.md [hunk 37]
```diff
@@ -0,0 +1,77 @@
...
+
+NAME: ...
```
Comment (nickboldt) [inline scripts/install.md:65]: are these empty lines really helpful? would sample content in there be more helpful than `...` ? what are you trying to tell the user here, just the list of metadata names, but not actually any sample data?

File: scripts/install.md [hunk 38]
```diff
@@ -0,0 +1,77 @@
+## **Documentation for the script to handle CLI helm installation for OCP**
```
Comment (nickboldt) [inline scripts/install.md:1]: why not just...

```suggestion
## **Helm CLI installation for OCP**
```

?

File: scripts/install.sh (hunk 30 above)
Comment (Fortune-Ndlovu) [reply scripts/install.sh:150]: I see to help with debugging

File: scripts/install.sh [hunk 39]
```diff
@@ -0,0 +1,151 @@
...
+        echo "Error: Namespace could not be detected. Please provide it using the --namespace flag."
+        exit 1
+    fi
+fi
+
+# Always include the router base in Helm arguments
+EXTRA_HELM_ARGS+=" --set global.clusterRouterBase=$ROUTER_BASE"
+
+# Construct Helm install command
+HELM_CMD="helm install"
```
Comment (nickboldt) [inline scripts/install.sh:130]: In case there's an existing chart installed, this will upgrade it; or if not installed, it will install it. Best of both worlds. 

```suggestion
HELM_CMD="helm upgrade -i"
```

**Review by rm3l — APPROVED** [review]
LGTM. Tested on Dev Sandbox and ClusterBot.

**Comment by nickboldt** [comment]
PR abandoned, because https://issues.redhat.com/browse/RHIDP-2706 was closed.
